thumbfast.lua

```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+        local thumb_file = io.open(filename .. ".bgra", "wb")
+        for line = 0, thumbnail_size.h - 1 do
+            atlas:seek("set", 4 * x_start + (y_start + line) * stride)
+            local data = atlas:read(thumbnail_size.w * 4)
+            if data ~= nil then
+                thumb_file:write(data)
+            end
+        end
+        thumb_file:close()
+        storyboard_thumbnails[thumb_idx] = filename
+        if last_seek_time then
+            local last_thumb_idx = math.floor(last_seek_time / thumbnail_delta)
+            if last_thumb_idx == thumb_idx then
+                last_seek_time = nil
+            end
+        end
+    end
+    atlas:close()
+end
+
+local function fetch_fragment(storyboard, i, thumbnail_size, storyboard_scale)
+    if not storyboard.fragments[i] then return end
+
+    local args = {
+        mpv_path, storyboard.fragments[i].url, "--no-config", "--msg-level=all=no", "--really-quiet", "--no-terminal",
+        "--frames=1",
+        --"--load-scripts=no", "--osc=no", "--ytdl=no", "--load-stats-overlay=no", "--load-osd-console=no", "--load-auto-profiles=no",
+        "--no-sub", "--no-audio", "--hr-seek=no", "--sub-font-provider=none", "--embeddedfonts=no",
+        "--no-ytdl", "--demuxer-readahead-secs=0", "--demuxer-max-bytes=128KiB",
+        "--vd-lavc-software-fallback=1", "--vd-lavc-fast", "--vd-lavc-threads=2", --"--hwdec="..(options.hwdec and "auto" or "no"),
+        --"--vf="..vf_string(filters_all, true),
+        "--sws-allow-zimg=no", "--sws-fast=yes", "--sws-scaler=fast-bilinear",
+        --"--video-rotate="..last_rotate,
+        "--vf-add=format=bgra,scale=round(iw*"..storyboard_scale.w.."/"..thumbnail_size.w..")*"..thumbnail_size.w..":round(ih*"..storyboard_scale.h.."/"..thumbnail_size.h..")*"..thumbnail_size.h,
+        "--ovc=rawvideo", "--of=rawvideo", "--ofopts=update=1", "--o="..options.thumbnail..".ytdl"..tostring(i)
+    }
+
+    if os_name == "Mac" then
+        table.insert(args, "--macos-app-activation-policy=prohibited")
+    end
+
+    spawn_queued_process(args,
+        function(success, result, err)
+            if success == false or (not result or result.status ~= 0) then
+                if not result.killed_by_us then
+                    mp.msg.error("thumbfast: storyboard download failed", "atlas:", i, "status:", result.status)
+                end
+            else
+                get_thumb(options.thumbnail..".ytdl"..tostring(i), i, storyboard, thumbnail_size, storyboard_scale)
+            end
+        end
+    )
+
+    fetch_fragment(storyboard, i+1, thumbnail_size, storyboard_scale)
+end
+
+local function anycase(s)
+    return string.gsub(s, "%a", function (c)
+        return string.format("[%s%s]", c:lower(), c:upper())
+    end)
+end
+
+local http_prefix = anycase("^https?://")
+local ytdl_prefix = "^ytdl://(.+)"
+local subdomains = "[%w-.]*"
+local naked_ytdl_id = "^ytdl://([%w-_]+)$"
+local youtube_id = "[%w-_]+.*"
```

Comment:
We know they are only ever 11 chars, and base64:
```suggestion
local youtube_id = "[%w-_]{11}.*"
```

Reply:
No such thing in Lua. Of course, we can repeat the character class 11 times instead. I'm just not too sure if there could be video IDs shorter than 11 characters that redirect to the canonical ID, I know it happens if you add garbage at the end of a valid 11-char video ID. This costs nothing and is safer (: yt-dlp handles too-long video IDs just fine